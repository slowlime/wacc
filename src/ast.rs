use crate::token::{Symbol, Token};

macro_rules! define_op_kind {
    ($name:ident { $( $op:ident => $symbol:ident, )+ }) => {
        #[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
        pub enum $name {
            $( $op, )+
        }

        impl $name {
            pub fn symbol(self) -> Symbol {
                match self {
                    $( Self::$op => Symbol::$symbol, )+
                }
            }
        }

        impl TryFrom<Symbol> for $name {
            type Error = ();

            fn try_from(symbol: Symbol) -> Result<Self, Self::Error> {
                match symbol {
                    $( Symbol::$symbol => Ok(Self::$op), )+
                    _ => Err(()),
                }
            }
        }
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Program<'a> {
    pub classes: Vec<Class<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Class<'a> {
    pub name: Name<'a>,
    pub inherits: Option<Name<'a>>,
    pub features: Vec<Feature<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Name<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Feature<'a> {
    Method(Method<'a>),
    Field(Binding<'a>),
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Method<'a> {
    pub name: Name<'a>,
    pub params: Vec<Formal<'a>>,
    pub return_ty: Name<'a>,
    pub body: Expr<'a>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Formal<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Binding<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
    pub init: Option<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Expr<'a> {
    Assignment(Assignment<'a>),
    Call(Call<'a>),
    If(If<'a>),
    While(While<'a>),
    Block(Block<'a>),
    Let(Let<'a>),
    Case(Case<'a>),
    New(New<'a>),
    BinOp(BinOpExpr<'a>),
    UnOp(UnOpExpr<'a>),
    Name(Name<'a>),
    Int(IntLit<'a>),
    String(StringLit<'a>),
    Bool(BoolLit<'a>),
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Assignment<'a> {
    pub name: Name<'a>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Call<'a> {
    pub receiver: Receiver<'a>,
    pub method: Name<'a>,
    pub args: Vec<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Receiver<'a> {
    SelfType,
    Dynamic(Box<Expr<'a>>),
    Static {
        object: Box<Expr<'a>>,
        ty: Name<'a>,
    },
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct If<'a> {
    pub condition: Box<Expr<'a>>,
    pub consequent: Box<Expr<'a>>,
    pub alternative: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct While<'a> {
    pub condition: Box<Expr<'a>>,
    pub body: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Block<'a> {
    pub body: Vec<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Let<'a> {
    pub bindings: Vec<Binding<'a>>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Case<'a> {
    pub scrutinee: Box<Expr<'a>>,
    pub arms: Vec<CaseArm<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct CaseArm<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct New<'a>(pub Name<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct BinOpExpr<'a> {
    pub op: BinOpKind,
    pub lhs: Box<Expr<'a>>,
    pub rhs: Box<Expr<'a>>,
}

define_op_kind!(BinOpKind {
    Add => Plus,
    Subtract => Minus,
    Multiply => Asterisk,
    Divide => Slash,
    LessThan => Less,
    LessEquals => LessEquals,
    Equals => Equals,
});

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct UnOpExpr<'a> {
    pub op: UnOpKind,
    pub expr: Box<Expr<'a>>,
}

define_op_kind!(UnOpKind {
    IsVoid => IsVoid,
    Complement => Tilde,
    Not => Not,
});

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct IntLit<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct StringLit<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct BoolLit<'a>(pub Token<'a>);
