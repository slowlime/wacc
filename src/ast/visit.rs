use super::*;

pub trait AstRecurse<'buf> {
    fn recurse<V: Visitor<'buf, Output = ()>>(&self, visitor: &mut V);
    fn recurse_mut<V: VisitorMut<'buf, Output = ()>>(&mut self, visitor: &mut V);
}

macro_rules! define_visitor {
    ($( $type:ident { $( $name:ident ( $arg:ident : $ty:ty ) );+ $(;)? } )+) => {
        pub trait Visitor<'buf>
        where
            Self: Sized,
        {
            type Output;

            $(
                $(
                    fn $name(&mut self, $arg: &$ty) -> Self::Output;
                )+
            )+
        }

        pub trait VisitorMut<'buf>
        where
            Self: Sized,
        {
            type Output;

            $(
                $(
                    fn $name(&mut self, $arg: &mut $ty) -> Self::Output;
                )+
            )+
        }

        pub trait DefaultVisitor<'buf>
        where
            Self: Sized,
        {
            $( define_visitor!(@ $type { $( $name ( $arg : &$ty ) => recurse; )+ } ); )+
        }

        impl<'buf, T> Visitor<'buf> for T
        where
            T: DefaultVisitor<'buf>,
        {
            type Output = ();

            $(
                $(
                    fn $name(&mut self, $arg: &$ty) {
                        <Self as DefaultVisitor<'buf>>::$name(self, $arg);
                    }
                )+
            )+
        }

        pub trait DefaultVisitorMut<'buf>
        where
            Self: Sized,
        {
            $( define_visitor!(@ $type { $( $name ( $arg : &mut $ty ) => recurse_mut; )+ } ); )+
        }

        impl<'buf, T> VisitorMut<'buf> for T
        where
            T: DefaultVisitorMut<'buf>,
        {
            type Output = ();

            $(
                $(
                    fn $name(&mut self, $arg: &mut $ty) {
                        <Self as DefaultVisitorMut<'buf>>::$name(self, $arg);
                    }
                )+
            )+
        }
    };

    (@ NonTerminal { $( $name:ident ( $arg:ident : $ty:ty ) => $recurse:ident; )+ }) => {
        $(
            fn $name(&mut self, $arg: $ty) {
                $arg.$recurse(self);
            }
        )+
    };

    (@ Terminal { $( $name:ident ( $arg:ident : $ty:ty ) => $recurse:ident; )+ }) => {
        $(
            #[allow(unused_variables)]
            fn $name(&mut self, $arg: $ty) {}
        )+
    };
}

define_visitor! {
    NonTerminal {
        // statements
        visit_program(program: Program<'buf>);
        visit_class(class: Class<'buf>);
        visit_feature(feature: Feature<'buf>);
        visit_method(method: Method<'buf>);
        visit_field(field: Field<'buf>);

        // expr
        visit_expr(expr: Expr<'buf>);
        visit_assignment(expr: Assignment<'buf>);
        visit_call(expr: Call<'buf>);
        visit_if(expr: If<'buf>);
        visit_while(expr: While<'buf>);
        visit_block(expr: Block<'buf>);
        visit_let(expr: Let<'buf>);
        visit_case(expr: Case<'buf>);
        visit_new(expr: New<'buf>);
        visit_bin_op(expr: BinOpExpr<'buf>);
        visit_un_op(expr: UnOpExpr<'buf>);
        visit_name_expr(expr: NameExpr<'buf>);

        // neither stmt nor expr
        visit_formal(formal: Formal<'buf>);
        visit_receiver(recv: Receiver<'buf>);
        visit_case_arm(arm: CaseArm<'buf>);
        visit_ty_name(ty_name: TyName<'buf>);
        visit_binding(binding: Binding<'buf>);
    }

    Terminal {
        visit_name(name: Name<'buf>);
        visit_int_lit(expr: IntLit);
        visit_string_lit(expr: StringLit<'buf>);
        visit_bool_lit(expr: BoolLit);
    }
}
