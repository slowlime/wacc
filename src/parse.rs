mod cursor;
mod lexer;
mod parser;
pub mod token;

pub use cursor::Cursor;
pub use lexer::{Lexer, LexerError};
pub use parser::{Parser, ParserError};
