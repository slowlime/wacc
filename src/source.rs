use std::fs;
use std::io;
use std::num::NonZeroUsize;
use std::path::{Path, PathBuf};

use elsa::FrozenVec;
use serde::Serialize;

#[derive(Serialize, Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct SourceId(NonZeroUsize);

#[derive(Default)]
pub struct SourceBuffer(FrozenVec<Vec<u8>>);

impl SourceBuffer {
    pub fn new() -> Self {
        Self(FrozenVec::new())
    }
}

pub struct Source<'buf> {
    buf: &'buf SourceBuffer,
    files: Vec<SourceFile<'buf>>,
}

#[derive(Debug)]
pub struct SourceFile<'buf> {
    id: SourceId,
    path: PathBuf,
    buf: &'buf [u8],
}

impl<'buf> Source<'buf> {
    pub fn new(buf: &'buf mut SourceBuffer) -> Self {
        // take a &mut to ensure exclusivity of the borrow

        Self {
            buf: &*buf,
            files: Vec::new(),
        }
    }

    pub fn load_from_string(&mut self, path: PathBuf, buf: Vec<u8>) -> SourceId {
        let idx = self.files.len();
        assert_eq!(self.buf.0.len(), idx);
        let id = SourceId((idx + 1).try_into().unwrap());

        self.buf.0.push(buf);
        let buf: &'buf [u8] = &self.buf.0[idx];

        self.files.push(SourceFile { id, path, buf });

        id
    }

    pub fn load(&mut self, path: PathBuf) -> io::Result<SourceId> {
        let buf = fs::read(&path)?;

        Ok(self.load_from_string(path, buf))
    }

    pub fn get(&self, id: SourceId) -> Option<&SourceFile<'buf>> {
        self.files.get(usize::from(id.0) - 1)
    }

    pub fn iter(&self) -> impl Iterator<Item = &SourceFile<'buf>> {
        self.files.iter()
    }
}

impl<'buf> SourceFile<'buf> {
    pub fn id(&self) -> SourceId {
        self.id
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    pub fn buf(&self) -> &'buf [u8] {
        self.buf
    }
}
