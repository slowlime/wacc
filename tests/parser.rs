use std::borrow::Cow;
use std::path::Path;

use insta::assert_ron_snapshot;
use once_cell::unsync::OnceCell;
use paste::paste;
use pretty_assertions::assert_str_eq;

use wacc::ast::dump::{dump_ast, AstDumpFormat};
use wacc::parse::{Cursor, Lexer, Parser};
use wacc::source::{Source, SourceBuffer};

macro_rules! run_parser_test {
    ($filename:expr) => ({
        let mut source_buf = SourceBuffer::new();
        let mut source = Source::new(&mut source_buf);
        let code = include_bytes!(concat!("parser/", $filename, ".cl"));
        let source_id = source.load_from_string(
            Path::new("tests").join("parser").join($filename.to_string() + ".cl"),
            code.to_vec()
        );
        let source_file = source.get(source_id).unwrap();

        let cursor = Cursor::new(source_file);
        let lexer = Lexer::new(cursor);
        let parser = Parser::new(lexer);

        let parse_result = parser.parse();

        insta::with_settings!({
            description => String::from_utf8_lossy(code),
        }, {
            assert_ron_snapshot!(parse_result);
        });

        struct Dump<'a> {
            bytes: Cow<'a, [u8]>,
            string: OnceCell<String>,
        }

        impl PartialEq for Dump<'_> {
            fn eq(&self, other: &Dump) -> bool {
                self.bytes == other.bytes
            }
        }

        impl AsRef<str> for Dump<'_> {
            fn as_ref(&self) -> &str {
                self.string.get_or_init(|| {
                    String::from_utf8_lossy(&self.bytes).into_owned()
                })
            }
        }

        impl<'a> From<&'a [u8]> for Dump<'a> {
            fn from(bytes: &'a [u8]) -> Self {
                Self {
                    bytes: Cow::Borrowed(bytes),
                    string: OnceCell::new(),
                }
            }
        }

        impl From<Vec<u8>> for Dump<'_> {
            fn from(bytes: Vec<u8>) -> Self {
                Self {
                    bytes: Cow::Owned(bytes),
                    string: OnceCell::new(),
                }
            }
        }

        let coolc_actual = Dump::from(match parse_result {
            Ok(ast) => {
                let mut buf = vec![];
                dump_ast(
                    &source,
                    AstDumpFormat::Coolc {
                        ignore_types: true,
                    },
                    ast,
                    &mut buf
                )
                .unwrap();

                buf
            },

            Err(_) => b"ERR parsing failed\n".to_vec(),
        });

        let coolc_expected = Dump::from(
            include_bytes!(concat!("parser/", $filename, ".coolc-parse")).as_slice()
        );

        assert_str_eq!(coolc_actual, coolc_expected);
    });

    { $( $filename:ident ),+ $(,)? } => {
        $(
            paste! {
                #[test]
                fn [< test_ $filename >]() {
                    run_parser_test!(concat!(stringify!($filename)));
                }
            }
        )+
    }
}

run_parser_test! {
    fail_empty,
    pass_simple_class,
    pass_arith,
    pass_multi_class,
    pass_fields,
    pass_priority_assign,
    pass_priority_arith,
}
