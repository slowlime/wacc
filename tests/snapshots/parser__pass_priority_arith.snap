---
source: tests/parser.rs
description: "class Main {\n    x : Int;\n    y : Int;\n    z : Int;\n\n    main(): Object {{\n        x - x < y + y;\n        x + y * ~z = x * y / y * z / z * x * new Int;\n    }};\n};\n"
expression: parse_result
---
Ok(Program(
  classes: [
    Class(
      name: TyName(Name(Spanned(
        value: "Main",
        span: Span(
          start: Position(
            src: Some(SourceId(1)),
            byte: 6,
            line: 1,
            col: 7,
          ),
          end: Position(
            src: Some(SourceId(1)),
            byte: 9,
            line: 1,
            col: 10,
          ),
        ),
      ))),
      inherits: None,
      features: [
        Field(Field(Binding(
          name: Name(Spanned(
            value: "x",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 17,
                line: 2,
                col: 5,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 17,
                line: 2,
                col: 5,
              ),
            ),
          )),
          binding_id: None,
          ty_name: TyName(Name(Spanned(
            value: "Int",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 21,
                line: 2,
                col: 9,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 23,
                line: 2,
                col: 11,
              ),
            ),
          ))),
          init: None,
          span: Span(
            start: Position(
              src: Some(SourceId(1)),
              byte: 17,
              line: 2,
              col: 5,
            ),
            end: Position(
              src: Some(SourceId(1)),
              byte: 17,
              line: 2,
              col: 5,
            ),
          ),
          ty: Unresolved(Named(TyName(Name(Spanned(
            value: "Int",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 21,
                line: 2,
                col: 9,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 23,
                line: 2,
                col: 11,
              ),
            ),
          ))))),
        ))),
        Field(Field(Binding(
          name: Name(Spanned(
            value: "y",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 30,
                line: 3,
                col: 5,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 30,
                line: 3,
                col: 5,
              ),
            ),
          )),
          binding_id: None,
          ty_name: TyName(Name(Spanned(
            value: "Int",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 34,
                line: 3,
                col: 9,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 36,
                line: 3,
                col: 11,
              ),
            ),
          ))),
          init: None,
          span: Span(
            start: Position(
              src: Some(SourceId(1)),
              byte: 30,
              line: 3,
              col: 5,
            ),
            end: Position(
              src: Some(SourceId(1)),
              byte: 30,
              line: 3,
              col: 5,
            ),
          ),
          ty: Unresolved(Named(TyName(Name(Spanned(
            value: "Int",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 34,
                line: 3,
                col: 9,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 36,
                line: 3,
                col: 11,
              ),
            ),
          ))))),
        ))),
        Field(Field(Binding(
          name: Name(Spanned(
            value: "z",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 43,
                line: 4,
                col: 5,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 43,
                line: 4,
                col: 5,
              ),
            ),
          )),
          binding_id: None,
          ty_name: TyName(Name(Spanned(
            value: "Int",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 47,
                line: 4,
                col: 9,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 49,
                line: 4,
                col: 11,
              ),
            ),
          ))),
          init: None,
          span: Span(
            start: Position(
              src: Some(SourceId(1)),
              byte: 43,
              line: 4,
              col: 5,
            ),
            end: Position(
              src: Some(SourceId(1)),
              byte: 43,
              line: 4,
              col: 5,
            ),
          ),
          ty: Unresolved(Named(TyName(Name(Spanned(
            value: "Int",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 47,
                line: 4,
                col: 9,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 49,
                line: 4,
                col: 11,
              ),
            ),
          ))))),
        ))),
        Method(Method(
          name: Name(Spanned(
            value: "main",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 57,
                line: 6,
                col: 5,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 60,
                line: 6,
                col: 8,
              ),
            ),
          )),
          params: [],
          return_ty: TyName(Name(Spanned(
            value: "Object",
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 65,
                line: 6,
                col: 13,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 70,
                line: 6,
                col: 18,
              ),
            ),
          ))),
          body: Block(Block(
            body: [
              BinOp(BinOpExpr(
                op: LessThan,
                lhs: BinOp(BinOpExpr(
                  op: Subtract,
                  lhs: Name(NameExpr(
                    name: Name(Spanned(
                      value: "x",
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 83,
                          line: 7,
                          col: 9,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 83,
                          line: 7,
                          col: 9,
                        ),
                      ),
                    )),
                    binding_id: None,
                    ty: None,
                  )),
                  rhs: Name(NameExpr(
                    name: Name(Spanned(
                      value: "x",
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 87,
                          line: 7,
                          col: 13,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 87,
                          line: 7,
                          col: 13,
                        ),
                      ),
                    )),
                    binding_id: None,
                    ty: None,
                  )),
                  span: Span(
                    start: Position(
                      src: Some(SourceId(1)),
                      byte: 83,
                      line: 7,
                      col: 9,
                    ),
                    end: Position(
                      src: Some(SourceId(1)),
                      byte: 83,
                      line: 7,
                      col: 9,
                    ),
                  ),
                  ty: None,
                )),
                rhs: BinOp(BinOpExpr(
                  op: Add,
                  lhs: Name(NameExpr(
                    name: Name(Spanned(
                      value: "y",
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 91,
                          line: 7,
                          col: 17,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 91,
                          line: 7,
                          col: 17,
                        ),
                      ),
                    )),
                    binding_id: None,
                    ty: None,
                  )),
                  rhs: Name(NameExpr(
                    name: Name(Spanned(
                      value: "y",
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 95,
                          line: 7,
                          col: 21,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 95,
                          line: 7,
                          col: 21,
                        ),
                      ),
                    )),
                    binding_id: None,
                    ty: None,
                  )),
                  span: Span(
                    start: Position(
                      src: Some(SourceId(1)),
                      byte: 91,
                      line: 7,
                      col: 17,
                    ),
                    end: Position(
                      src: Some(SourceId(1)),
                      byte: 91,
                      line: 7,
                      col: 17,
                    ),
                  ),
                  ty: None,
                )),
                span: Span(
                  start: Position(
                    src: Some(SourceId(1)),
                    byte: 83,
                    line: 7,
                    col: 9,
                  ),
                  end: Position(
                    src: Some(SourceId(1)),
                    byte: 83,
                    line: 7,
                    col: 9,
                  ),
                ),
                ty: None,
              )),
              BinOp(BinOpExpr(
                op: Equals,
                lhs: BinOp(BinOpExpr(
                  op: Add,
                  lhs: Name(NameExpr(
                    name: Name(Spanned(
                      value: "x",
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 106,
                          line: 8,
                          col: 9,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 106,
                          line: 8,
                          col: 9,
                        ),
                      ),
                    )),
                    binding_id: None,
                    ty: None,
                  )),
                  rhs: BinOp(BinOpExpr(
                    op: Multiply,
                    lhs: Name(NameExpr(
                      name: Name(Spanned(
                        value: "y",
                        span: Span(
                          start: Position(
                            src: Some(SourceId(1)),
                            byte: 110,
                            line: 8,
                            col: 13,
                          ),
                          end: Position(
                            src: Some(SourceId(1)),
                            byte: 110,
                            line: 8,
                            col: 13,
                          ),
                        ),
                      )),
                      binding_id: None,
                      ty: None,
                    )),
                    rhs: UnOp(UnOpExpr(
                      op: Complement,
                      expr: Name(NameExpr(
                        name: Name(Spanned(
                          value: "z",
                          span: Span(
                            start: Position(
                              src: Some(SourceId(1)),
                              byte: 115,
                              line: 8,
                              col: 18,
                            ),
                            end: Position(
                              src: Some(SourceId(1)),
                              byte: 115,
                              line: 8,
                              col: 18,
                            ),
                          ),
                        )),
                        binding_id: None,
                        ty: None,
                      )),
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 114,
                          line: 8,
                          col: 17,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 114,
                          line: 8,
                          col: 17,
                        ),
                      ),
                      ty: None,
                    )),
                    span: Span(
                      start: Position(
                        src: Some(SourceId(1)),
                        byte: 110,
                        line: 8,
                        col: 13,
                      ),
                      end: Position(
                        src: Some(SourceId(1)),
                        byte: 110,
                        line: 8,
                        col: 13,
                      ),
                    ),
                    ty: None,
                  )),
                  span: Span(
                    start: Position(
                      src: Some(SourceId(1)),
                      byte: 106,
                      line: 8,
                      col: 9,
                    ),
                    end: Position(
                      src: Some(SourceId(1)),
                      byte: 106,
                      line: 8,
                      col: 9,
                    ),
                  ),
                  ty: None,
                )),
                rhs: BinOp(BinOpExpr(
                  op: Multiply,
                  lhs: BinOp(BinOpExpr(
                    op: Multiply,
                    lhs: BinOp(BinOpExpr(
                      op: Divide,
                      lhs: BinOp(BinOpExpr(
                        op: Multiply,
                        lhs: BinOp(BinOpExpr(
                          op: Divide,
                          lhs: BinOp(BinOpExpr(
                            op: Multiply,
                            lhs: Name(NameExpr(
                              name: Name(Spanned(
                                value: "x",
                                span: Span(
                                  start: Position(
                                    src: Some(SourceId(1)),
                                    byte: 119,
                                    line: 8,
                                    col: 22,
                                  ),
                                  end: Position(
                                    src: Some(SourceId(1)),
                                    byte: 119,
                                    line: 8,
                                    col: 22,
                                  ),
                                ),
                              )),
                              binding_id: None,
                              ty: None,
                            )),
                            rhs: Name(NameExpr(
                              name: Name(Spanned(
                                value: "y",
                                span: Span(
                                  start: Position(
                                    src: Some(SourceId(1)),
                                    byte: 123,
                                    line: 8,
                                    col: 26,
                                  ),
                                  end: Position(
                                    src: Some(SourceId(1)),
                                    byte: 123,
                                    line: 8,
                                    col: 26,
                                  ),
                                ),
                              )),
                              binding_id: None,
                              ty: None,
                            )),
                            span: Span(
                              start: Position(
                                src: Some(SourceId(1)),
                                byte: 119,
                                line: 8,
                                col: 22,
                              ),
                              end: Position(
                                src: Some(SourceId(1)),
                                byte: 119,
                                line: 8,
                                col: 22,
                              ),
                            ),
                            ty: None,
                          )),
                          rhs: Name(NameExpr(
                            name: Name(Spanned(
                              value: "y",
                              span: Span(
                                start: Position(
                                  src: Some(SourceId(1)),
                                  byte: 127,
                                  line: 8,
                                  col: 30,
                                ),
                                end: Position(
                                  src: Some(SourceId(1)),
                                  byte: 127,
                                  line: 8,
                                  col: 30,
                                ),
                              ),
                            )),
                            binding_id: None,
                            ty: None,
                          )),
                          span: Span(
                            start: Position(
                              src: Some(SourceId(1)),
                              byte: 119,
                              line: 8,
                              col: 22,
                            ),
                            end: Position(
                              src: Some(SourceId(1)),
                              byte: 119,
                              line: 8,
                              col: 22,
                            ),
                          ),
                          ty: None,
                        )),
                        rhs: Name(NameExpr(
                          name: Name(Spanned(
                            value: "z",
                            span: Span(
                              start: Position(
                                src: Some(SourceId(1)),
                                byte: 131,
                                line: 8,
                                col: 34,
                              ),
                              end: Position(
                                src: Some(SourceId(1)),
                                byte: 131,
                                line: 8,
                                col: 34,
                              ),
                            ),
                          )),
                          binding_id: None,
                          ty: None,
                        )),
                        span: Span(
                          start: Position(
                            src: Some(SourceId(1)),
                            byte: 119,
                            line: 8,
                            col: 22,
                          ),
                          end: Position(
                            src: Some(SourceId(1)),
                            byte: 119,
                            line: 8,
                            col: 22,
                          ),
                        ),
                        ty: None,
                      )),
                      rhs: Name(NameExpr(
                        name: Name(Spanned(
                          value: "z",
                          span: Span(
                            start: Position(
                              src: Some(SourceId(1)),
                              byte: 135,
                              line: 8,
                              col: 38,
                            ),
                            end: Position(
                              src: Some(SourceId(1)),
                              byte: 135,
                              line: 8,
                              col: 38,
                            ),
                          ),
                        )),
                        binding_id: None,
                        ty: None,
                      )),
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 119,
                          line: 8,
                          col: 22,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 119,
                          line: 8,
                          col: 22,
                        ),
                      ),
                      ty: None,
                    )),
                    rhs: Name(NameExpr(
                      name: Name(Spanned(
                        value: "x",
                        span: Span(
                          start: Position(
                            src: Some(SourceId(1)),
                            byte: 139,
                            line: 8,
                            col: 42,
                          ),
                          end: Position(
                            src: Some(SourceId(1)),
                            byte: 139,
                            line: 8,
                            col: 42,
                          ),
                        ),
                      )),
                      binding_id: None,
                      ty: None,
                    )),
                    span: Span(
                      start: Position(
                        src: Some(SourceId(1)),
                        byte: 119,
                        line: 8,
                        col: 22,
                      ),
                      end: Position(
                        src: Some(SourceId(1)),
                        byte: 119,
                        line: 8,
                        col: 22,
                      ),
                    ),
                    ty: None,
                  )),
                  rhs: New(New(
                    ty_name: TyName(Name(Spanned(
                      value: "Int",
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 147,
                          line: 8,
                          col: 50,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 149,
                          line: 8,
                          col: 52,
                        ),
                      ),
                    ))),
                    span: Span(
                      start: Position(
                        src: Some(SourceId(1)),
                        byte: 143,
                        line: 8,
                        col: 46,
                      ),
                      end: Position(
                        src: Some(SourceId(1)),
                        byte: 145,
                        line: 8,
                        col: 48,
                      ),
                    ),
                    ty: Unresolved(Named(TyName(Name(Spanned(
                      value: "Int",
                      span: Span(
                        start: Position(
                          src: Some(SourceId(1)),
                          byte: 147,
                          line: 8,
                          col: 50,
                        ),
                        end: Position(
                          src: Some(SourceId(1)),
                          byte: 149,
                          line: 8,
                          col: 52,
                        ),
                      ),
                    ))))),
                  )),
                  span: Span(
                    start: Position(
                      src: Some(SourceId(1)),
                      byte: 119,
                      line: 8,
                      col: 22,
                    ),
                    end: Position(
                      src: Some(SourceId(1)),
                      byte: 119,
                      line: 8,
                      col: 22,
                    ),
                  ),
                  ty: None,
                )),
                span: Span(
                  start: Position(
                    src: Some(SourceId(1)),
                    byte: 106,
                    line: 8,
                    col: 9,
                  ),
                  end: Position(
                    src: Some(SourceId(1)),
                    byte: 106,
                    line: 8,
                    col: 9,
                  ),
                ),
                ty: None,
              )),
            ],
            span: Span(
              start: Position(
                src: Some(SourceId(1)),
                byte: 73,
                line: 6,
                col: 21,
              ),
              end: Position(
                src: Some(SourceId(1)),
                byte: 73,
                line: 6,
                col: 21,
              ),
            ),
          )),
          span: Span(
            start: Position(
              src: Some(SourceId(1)),
              byte: 57,
              line: 6,
              col: 5,
            ),
            end: Position(
              src: Some(SourceId(1)),
              byte: 60,
              line: 6,
              col: 8,
            ),
          ),
          ty: Unresolved(Function(
            args: [],
            ret: Unresolved(Named(TyName(Name(Spanned(
              value: "Object",
              span: Span(
                start: Position(
                  src: Some(SourceId(1)),
                  byte: 65,
                  line: 6,
                  col: 13,
                ),
                end: Position(
                  src: Some(SourceId(1)),
                  byte: 70,
                  line: 6,
                  col: 18,
                ),
              ),
            ))))),
          )),
        )),
      ],
      span: Span(
        start: Position(
          src: Some(SourceId(1)),
          byte: 0,
          line: 1,
          col: 1,
        ),
        end: Position(
          src: Some(SourceId(1)),
          byte: 4,
          line: 1,
          col: 5,
        ),
      ),
      ty: Unresolved(Named(TyName(Name(Spanned(
        value: "Main",
        span: Span(
          start: Position(
            src: Some(SourceId(1)),
            byte: 6,
            line: 1,
            col: 7,
          ),
          end: Position(
            src: Some(SourceId(1)),
            byte: 9,
            line: 1,
            col: 10,
          ),
        ),
      ))))),
    ),
  ],
  span: Span(
    start: Position(
      src: Some(SourceId(1)),
      byte: 0,
      line: 1,
      col: 1,
    ),
    end: Position(
      src: Some(SourceId(1)),
      byte: 4,
      line: 1,
      col: 5,
    ),
  ),
))
